File: LargestSubstringBetweenTwoEqualCharacters.py
# Given a string s, return the length of the longest substring between two equal characters, excluding the two characters. If there is no such substring return -1.
# A substring is a contiguous sequence of characters within a string.

def maxLengthBetweenEqualCharacters(s):
    maxLen = -1
    p1 = len(s) - 1

    while p1 >= 0:
        found = s.find(s[p1], 0, p1)
        if found != -1:
            tempLen = (p1 - found) - 1
            if  tempLen > maxLen:
                maxLen = tempLen
        p1 -= 1
    return maxLen

#Test Cases
s = "aa"
print(maxLengthBetweenEqualCharacters(s))

s = "abca"
print(maxLengthBetweenEqualCharacters(s))

s = "cbzxy"
print(maxLengthBetweenEqualCharacters(s))

s = "cbzcxy"
print(maxLengthBetweenEqualCharacters(s))

s = "bczcxy"
print(maxLengthBetweenEqualCharacters(s))

s = "cabbac"
print(maxLengthBetweenEqualCharacters(s))

s = "mgntdygtxrvxjnwksqhxuxtrv"
print(maxLengthBetweenEqualCharacters(s))

File: LargestSubstringBetweenTwoEqualCharacters.js
// Given a string s, return the length of the longest substring between two equal characters, excluding the two characters. If there is no such substring return -1.
// A substring is a contiguous sequence of characters within a string.

function maxLengthBetweenEqualCharacter(s){
    charIndx = {}
    maxLen = -1

    for(i = 0; i < s.length; i++){
        if (s[i] in charIndx){
            charIndx[s[i]].push(i)
        }else{
            charIndx[s[i]] = [i]
        }
    }

    for (value of Object.values(charIndx)){
        maxLen = Math.max(value[value.length - 1] - value[0] - 1, maxLen)
    }

    return maxLen
}; 

// Test cases
s = "aa"
console.log(maxLengthBetweenEqualCharacter(s))

s = "abca"
console.log(maxLengthBetweenEqualCharacter(s))

s = "cbzxy"
console.log(maxLengthBetweenEqualCharacter(s))
